janus_core/helpers/janus_types.py

```diff
@@ -179,3 +131,122 @@ class EoSResults(TypedDict, total=False):
     bulk_modulus: float
     v_0: float
     e_0: float
+
+
+# pylint: disable=too-few-public-methods
+class Observable:
+    """
+    Observable data that may be correlated.
+
+    Parameters
+    ----------
+    dimension : int
+        The dimension of the observed data.
+    getter : Optional[callable]
+        An optional callable to construct the Observable from.
+    """
+
+    def __init__(self, dimension: int = 1, *, getter: Optional[callable] = None):
+        """
+        Initialise an observable with a given dimensionality.
+
+        Parameters
+        ----------
+        dimension : int
+            The dimension of the observed data.
+        getter : Optional[callable]
+            An optional callable to construct the Observable from.
+        """
+        self._dimension = dimension
+        self._getter = getter
+        self.atoms = None
+
+    def __call__(self, atoms: Atoms, *args, **kwargs) -> list[float]:
+        """
+        Call the user supplied getter if it exits.
+
+        Parameters
+        ----------
+        atoms : Atoms
+            Atoms object to extract values from.
+        *args : tuple
+            Additional positional arguments passed to getter.
+        **kwargs : dict
+            Additional kwargs passed getter.
+
+        Returns
+        -------
+        list[float]
+            The observed value, with dimensions atoms by self.dimension.
+
+        Raises
+        ------
+        ValueError
+            If user supplied getter is None.
+        """
+        if self._getter:
+            value = self._getter(atoms, *args, **kwargs)
+            if not isinstance(value, list):
+                return [value]
+            return value
+        raise ValueError("No user getter supplied")
+
+    @property
+    def dimension(self):
+        """
+        Dimension of the observable. Commensurate with self.__call__.
+
+        Returns
+        -------
+        int
+            Observables dimension.
+        """
+        return self._dimension
+
+    def atom_count(self, n_atoms: int):
+        """
+        Atom count to average over.
+
+        Parameters
+        ----------
+        n_atoms : int
+            Total possible atoms.
+
+        Returns
+        -------
+        int
+            Atom count averaged over.
+        """
+        if self.atoms:
+            if isinstance(self.atoms, list):
+                return len(self.atoms)
+            if isinstance(self.atoms, int):
+                return 1
+
+            start = self.atoms.start
+            stop = self.atoms.stop
+            step = self.atoms.step
```

Comment:
I'm probably missing something, but where are these attributes from?

Reply:
See above, this is the "mask" application. Currently only ```Velocity``` actually has a non-none value possible for ```self.atoms```. It could be a general kwarg for Observable, but then why would ```Stress``` allow atoms to be passed? Perhaps if you want stress computed over a subset of atoms, but I am unsure if that is doable easily in ase.

Reply:
> See above, this is the "mask" application. Currently only `Velocity` actually has a non-none value possible for `self.atoms`. 

I see, thanks. This is `atoms : Union[list[int], SliceLike]`, right?

One minor note is that I'd generally assume `atoms` would be an atoms object/list, so we could consider a less ambiguous name.

More significantly, lists and tuples don't have built-in `start`, `stop` or `step` attributes. Is this handled somewhere, or will this throw errors?

> It could be a general kwarg for Observable, but then why would `Stress` allow atoms to be passed? Perhaps if you want stress computed over a subset of atoms, but I am unsure if that is doable easily in ase.

I can imagine in general the ability to calculate any observable for an arbitrary subset could be useful, but if it's not trivial I think it makes sense to keep it case by case for now, as and when we have a useful application.